### PR TITLE
issue: 1521601 Allow user to change the flow tag id used in the steering

### PR DESCRIPTION
**Pull request by rafiw**

### CODE REVIEW

**Comment by liranoz12** [comment]
@rafiw Please fix conflicts.

File: src/vma/dev/ring_simple.cpp [hunk 1]
```diff
@@ -360,34 +360,24 @@ bool ring_simple::attach_flow(flow_tuple& flow_spec_5t, pkt_rcvr_sink *sink)
...
-					    flow_tag_id, FLOW_TAG_MASK);
-				flow_tag_id = FLOW_TAG_MASK;
-			}
-			ring_logdbg("sock_fd:%d enabled:%d with id:%d",
-				    flow_tag_id_candidate-1, m_flow_tag_enabled, flow_tag_id);
-		} else {
+		if (!flow_tag_id) {
 			flow_tag_id = FLOW_TAG_MASK; // FLOW_TAG_MASK - modal, FT to be attached but will not be used
-			ring_logdbg("flow_tag:%d disabled as flow_tag_id_candidate:%d", flow_tag_id, flow_tag_id_candidate);
+			ring_logdbg("flow_tag:%d disabled", flow_tag_id);
```
Comment (igor-ivanov) [inline src/vma/dev/ring_simple.cpp]: I think it is useless code as far as `flow_tag_id` is defined on sockinfo layer and here it should be guaranteed as >0 or FLOW_TAG_MASK.

File: src/vma/sock/sockinfo.h [hunk 2]
```diff
@@ -123,11 +123,13 @@ class sockinfo : public socket_fd_api, public pkt_rcvr_sink, public pkt_sndr_sou
...
 	inline void set_tcp_flow_is_5t(void) { m_tcp_flow_is_5t = true; }
-	inline void set_flow_tag(int flow_tag_id) {
-		if ( flow_tag_id && (flow_tag_id != FLOW_TAG_MASK)) {
+	inline bool set_flow_tag(uint32_t flow_tag_id) {
+		if (flow_tag_id && (flow_tag_id != FLOW_TAG_MASK)) {
 			m_flow_tag_id = flow_tag_id;
 			m_flow_tag_enabled = true;
+			return true;
 		}
+		return false;
```
Comment (igor-ivanov) [inline src/vma/sock/sockinfo.h]: I think that this code does not take into account that flow_tag_id should be limited by 20 bit (FLOW_TAG_MASK).
flow_tag_id can not exceed FLOW_TAG_MASK

File: src/vma/vma_extra.h [hunk 3]
```diff
@@ -51,6 +51,8 @@
 #define SO_VMA_USER_DATA	2801
 #define SO_VMA_RING_ALLOC_LOGIC 2810
 #define SO_VMA_RING_USER_MEMORY 2811
+#define SO_VMA_FLOW_TAG		2820
```
Comment (liranoz12) [inline src/vma/vma_extra.h]: Fix indentation

File: src/vma/sock/sockinfo.h [hunk 4]
```diff
@@ -210,6 +213,7 @@ class sockinfo : public socket_fd_api, public pkt_rcvr_sink, public pkt_sndr_sou
 	bool			m_tcp_flow_is_5t; // to bypass packet analysis
 
 	int*			m_p_rings_fds;
+	bool					m_user_def_flow_tag;
```
Comment (liranoz12) [inline src/vma/sock/sockinfo.h]: Fix indentation

File: src/vma/sock/sockinfo.cpp [hunk 5]
```diff
@@ -264,6 +266,31 @@ int sockinfo::setsockopt(int __level, int __optname, const void *__optval, sockl
...
+						ret = -1;
+						errno = EINVAL;
+					} else {
+						m_user_def_flow_tag = true;
+						si_logdbg("SO_VMA_FLOW_TAG, set "
+							  "socket %s to flow id %d",
+							  m_fd, m_flow_tag_id);
+					}
+				} else {
+					si_logdbg("SO_VMA_FLOW_TAG, bad length "
```
Comment (liranoz12) [inline src/vma/sock/sockinfo.cpp]: errno = EINVAL ?

File: src/vma/sock/sockinfo.cpp [hunk 6]
```diff
@@ -264,6 +266,31 @@ int sockinfo::setsockopt(int __level, int __optname, const void *__optval, sockl
...
+			}
+			break;
+		case SO_VMA_FLOW_TAG:
+			if (__optval) {
+				if (__optlen == sizeof(uint32_t)) {
+					if (!set_flow_tag(*(uint32_t*)__optval)) {
+						ret = -1;
+						errno = EINVAL;
+					} else {
+						m_user_def_flow_tag = true;
```
Comment (liranoz12) [inline src/vma/sock/sockinfo.cpp]: Ret should be updated to 0

File: src/vma/sock/sockinfo.cpp [hunk 7]
```diff
@@ -264,6 +266,31 @@ int sockinfo::setsockopt(int __level, int __optname, const void *__optval, sockl
...
 				errno = EINVAL;
 				si_logdbg("SOL_SOCKET, SO_VMA_RING_USER_MEMORY - NOT HANDLED, optval == NULL");
+
+			}
+			break;
+		case SO_VMA_FLOW_TAG:
+			if (__optval) {
+				if (__optlen == sizeof(uint32_t)) {
+					if (!set_flow_tag(*(uint32_t*)__optval)) {
+						ret = -1;
```
Comment (liranoz12) [inline src/vma/sock/sockinfo.cpp]: -1 is the default value

File: src/vma/dev/ring_simple.cpp [hunk 8]
```diff
@@ -504,7 +480,7 @@ bool ring_simple::attach_flow(flow_tuple& flow_spec_5t, pkt_rcvr_sink *sink)
 			m_lock_ring_rx.unlock();
 			if (safe_mce_sys().tcp_3t_rules) {
 				flow_tuple tcp_3t_only(flow_spec_5t.get_dst_ip(), flow_spec_5t.get_dst_port(), 0, 0, flow_spec_5t.get_protocol());
-				tcp_dst_port_filter = new rfs_rule_filter(m_tcp_dst_port_attach_map, rule_key.key, tcp_3t_only);
+				tcp_dst_port_filter = new (std::nothrow)rfs_rule_filter(m_tcp_dst_port_attach_map, rule_key.key, tcp_3t_only);
```
Comment (liranoz12) [inline src/vma/dev/ring_simple.cpp]: Please remove if it is redundant

File: src/vma/dev/ring_simple.cpp [hunk 9]
```diff
@@ -360,36 +360,14 @@ bool ring_simple::attach_flow(flow_tuple& flow_spec_5t, pkt_rcvr_sink *sink)
...
-				    flow_tag_id_candidate-1, m_flow_tag_enabled, flow_tag_id);
-		} else {
-			flow_tag_id = FLOW_TAG_MASK; // FLOW_TAG_MASK - modal, FT to be attached but will not be used
-			ring_logdbg("flow_tag:%d disabled as flow_tag_id_candidate:%d", flow_tag_id, flow_tag_id_candidate);
-		}
-	}
+	uint32_t flow_tag_id = si->get_flow_tag_val(); // spec will not be attached to rule
+
+	ring_logdbg("flow: %s, with sink (%p), m_flow_tag_enabled: %d",
+		    flow_spec_5t.to_str(), si, m_flow_tag_enabled);
```
Comment (liranoz12) [inline src/vma/dev/ring_simple.cpp]: Please add actual flow-tag value

File: src/vma/sock/sockinfo.h (hunk 2 above)
Comment (rafiw) [reply src/vma/sock/sockinfo.h]: done

File: src/vma/vma_extra.h (hunk 3 above)
Comment (rafiw) [reply src/vma/vma_extra.h]: done

File: src/vma/dev/ring_simple.cpp (hunk 8 above)
Comment (rafiw) [reply src/vma/dev/ring_simple.cpp]: done

File: src/vma/dev/ring_simple.cpp (hunk 9 above)
Comment (rafiw) [reply src/vma/dev/ring_simple.cpp]: done

**Comment by liranoz12** [comment]
bot:retest

**Comment by liranoz12** [comment]
bot:retest

**Pull request merged**